### PR TITLE
[DO-NOT-MERGE] SONARIAC-1305 Load dependency information based on "charts/" directory

### PR DESCRIPTION
**Pull request by mstachniuk**

### CODE REVIEW

File: iac-extensions/kubernetes/src/main/java/org/sonar/iac/kubernetes/plugin/HelmProcessor.java [hunk 1]
```diff
@@ -90,19 +93,58 @@ private static Map<String, String> validateAndReadFiles(InputFile inputFile, Map
...
     return fileContents;
   }
 
+  // exposed for tests
+  static Map<String, String> readCompressedFile(String name, InputFile inputFile) {
+    LOG.debug("Read dependency chart {}", name);
+    var pathPrefix = pathPrefix(name);
+    Map<String, String> result = new HashMap<>();
+    try (var source = inputFile.inputStream();
+      var gzip = new GZIPInputStream(source);
```
Comment (nils-werner-sonarsource) [inline iac-extensions/kubernetes/src/main/java/org/sonar/iac/kubernetes/plugin/HelmProcessor.java:112]: What is your security strategy to prevent vulnerabilities like zip slip? What about DOS? For me, this approach is super risky and I am not happy with handling compressed files. Please check with @agigleux if this is the approach we would like to have in the analyzer.

**Review by nils-werner-sonarsource — COMMENTED** [review]
Implementiationwise it is good, but I wonder if extracting a compressed file is secure.

File: iac-extensions/kubernetes/src/main/java/org/sonar/iac/helm/utils/HelmFilesystemUtils.java [hunk 2]
```diff
@@ -111,4 +108,31 @@ private static String resolveToInputFile(Path helmDirectoryPath, InputFile addit
   public static String normalizeToUnixPathSeparator(String filename) {
     return filename.replace('\\', '/');
   }
+
+  public static String normalizeToRuntimePathSeparator(String filename) {
+    return filename.replace('\\', File.separatorChar)
```
[the remark this answers is not in the document]
Comment (jonas-wielage-sonarsource) [reply iac-extensions/kubernetes/src/main/java/org/sonar/iac/helm/utils/HelmFilesystemUtils.java:113]: Additional comment: Also, maybe it would be easier to just use unix separators in the whole `HelmProcessor.readCompressedFile` method. Windows is able to read these paths as well.

**Review by jonas-wielage-sonarsource — COMMENTED** [review]
I don't see any changes related to loading dependency information from chart.yaml (title of this PR) maybe split the ticket?

File: iac-extensions/kubernetes/src/main/java/org/sonar/iac/helm/utils/HelmFilesystemUtils.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mstachniuk) [reply iac-extensions/kubernetes/src/main/java/org/sonar/iac/helm/utils/HelmFilesystemUtils.java:113]: I wanted to be sure if calling `File.getCanonicalPath()` works on each system. 
If `/` works everywhere then I'm ok to use it, but this requires additional tests.

File: iac-extensions/kubernetes/src/main/java/org/sonar/iac/helm/utils/HelmFilesystemUtils.java [hunk 3]
```diff
@@ -111,4 +108,31 @@ private static String resolveToInputFile(Path helmDirectoryPath, InputFile addit
...
+  }
+
+  public static boolean includeFile(String name) {
+    if (INCLUDED_EXTENSIONS.stream().anyMatch(name::endsWith)) {
+      var parent = new File(".");
+      var child = new File(parent, name);
+
+      boolean isPathSecure;
+      try {
+        isPathSecure = child.getCanonicalPath().startsWith(parent.getCanonicalPath() + File.separator) &&
```
[the remark this answers is not in the document]
Comment (mstachniuk) [reply iac-extensions/kubernetes/src/main/java/org/sonar/iac/helm/utils/HelmFilesystemUtils.java:124]: What if you have a directory named `aa..` or `..bb`? OK, it's a rare case similar to `~/file.txt` or `.` (current directory).
I just found that similar code mitigates the attack, because `File.getCanonicalPath()` returns absolute path without `..` and others and it's system-depended.

[the remark this answers is not in the document]
Comment (mstachniuk) [reply iac-extensions/kubernetes/src/main/java/org/sonar/iac/helm/utils/HelmFilesystemUtils.java:124]: If you still think that simple looking for `../`, `/..` and others is better I'm ok to change it.

File: iac-extensions/kubernetes/src/main/java/org/sonar/iac/kubernetes/plugin/HelmProcessor.java [hunk 4]
```diff
@@ -90,19 +96,58 @@ private static Map<String, String> validateAndReadFiles(InputFile inputFile, Map
...
+      var tar = new TarArchiveInputStream(gzip)) {
+      TarArchiveEntry entry;
+      while ((entry = tar.getNextEntry()) != null) {
+        var normalizedName = normalizeToRuntimePathSeparator(entry.getName());
+        if (entry.isFile() && HelmFilesystemUtils.includeFile(normalizedName)) {
+          var bytes = tar.readAllBytes();
```
[the remark this answers is not in the document]
Comment (mstachniuk) [reply iac-extensions/kubernetes/src/main/java/org/sonar/iac/kubernetes/plugin/HelmProcessor.java:121]: Yes it's misleading but I didn't design this API 🤷‍♂️

File: iac-extensions/kubernetes/src/test/java/org/sonar/iac/kubernetes/plugin/HelmProcessorTest.java [hunk 5]
```diff
@@ -227,10 +235,91 @@ void shouldNotEvaluateIfTemplateIsEmpty() throws IOException {
...
+    assertThat(logTester.logs(Level.DEBUG).get(1))
+      .startsWith("The path")
+      .endsWith("in compressed file looks suspicious, ignoring the file");
+  }
+
+  @Test
+  void shouldReadCompressedFileFilenameWithoutDirectory() throws IOException {
+    var inputFile = inputFile("src/test/resources/helm/charts/common-0.0.1.tgz");
+    var fileContents = HelmProcessor.readCompressedFile("common-0.0.1.tgz", inputFile);
+    var expectedKey = "common-0.0.1.tgz/common/values.yaml".replace('/', File.separatorChar);
```
[the remark this answers is not in the document]
Comment (mstachniuk) [reply iac-extensions/kubernetes/src/test/java/org/sonar/iac/kubernetes/plugin/HelmProcessorTest.java:283]: Actually this test is for coverage the following method:
```
  private static String pathPrefix(String name) {
    var index = name.lastIndexOf("/");
    if (index == -1) {
      return name;
    }
    return name.substring(0, index);
  }
```
In real-life cases there will be always `charts/some-filename.tgz` as first argument to `readCompressedFile()`

File: iac-extensions/kubernetes/src/test/resources/helm/charts/evil1-unix.tar.gz
[the remark this answers is not in the document]
Comment (mstachniuk) [reply iac-extensions/kubernetes/src/test/resources/helm/charts/evil1-unix.tar.gz:1]: The standard tools already fixed Zip Slip vulnerability, so this is why you don't see it

File: its/ruling/src/integrationTest/resources/expected/kubernetes/kubernetes-S2260.json [hunk 6]
```diff
@@ -2,6 +2,9 @@
 "kubernetes-project:ruling/src/integrationTest/resources/sources/kubernetes/S2260/parsing-error.yaml": [
 6
 ],
+"kubernetes-project:ruling/src/integrationTest/resources/sources/kubernetes/helm/chart-template-arch-dep/templates/service.yaml": [
+0
```
[the remark this answers is not in the document]
Comment (mstachniuk) [reply its/ruling/src/integrationTest/resources/expected/kubernetes/kubernetes-S2260.json:6]: Yes

File: iac-extensions/kubernetes/src/main/java/org/sonar/iac/helm/utils/HelmFilesystemUtils.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonas-wielage-sonarsource) [reply iac-extensions/kubernetes/src/main/java/org/sonar/iac/helm/utils/HelmFilesystemUtils.java:124]: I'm not sure it's better, I wasn't sure about the added benefit. But if it's an added layer of security and removes edge cases were unsure of, I think it's ok to use it.

File: iac-extensions/kubernetes/src/test/java/org/sonar/iac/kubernetes/plugin/HelmProcessorTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jonas-wielage-sonarsource) [reply iac-extensions/kubernetes/src/test/java/org/sonar/iac/kubernetes/plugin/HelmProcessorTest.java:283]: Ah ok, and than we want to have the `projectFolder/charts/someFileInAnArchive.yaml` path for the go Binary.

**Comment by mstachniuk** [comment]
The work on it is on hold for now.

File: iac-extensions/kubernetes/src/main/java/org/sonar/iac/kubernetes/plugin/HelmProcessor.java [hunk 7]
```diff
@@ -90,19 +96,62 @@ private static Map<String, String> validateAndReadFiles(InputFile inputFile, Map
...
```
Comment (mstachniuk) [inline iac-extensions/kubernetes/src/main/java/org/sonar/iac/kubernetes/plugin/HelmProcessor.java:126]: There should be protection in how many bytes should be read to avoid high memory consumption.

File: iac-extensions/kubernetes/src/main/java/org/sonar/iac/kubernetes/plugin/HelmProcessor.java [hunk 8]
```diff
@@ -90,19 +96,62 @@ private static Map<String, String> validateAndReadFiles(InputFile inputFile, Map
...
+    // The files are filter-out before, so it is enough to check only gz suffix, it will match *.tgz and *.gz files
+    return filenameToInputFile.getKey().endsWith("gz");
+  }
+
+  // exposed for tests
+  static Map<String, String> readCompressedFile(String name, InputFile inputFile) {
+    LOG.debug("Read dependency chart {}", name);
+    var pathPrefix = pathPrefix(name);
+    Map<String, String> result = new HashMap<>();
+    try (var source = inputFile.inputStream();
```
Comment (mstachniuk) [inline iac-extensions/kubernetes/src/main/java/org/sonar/iac/kubernetes/plugin/HelmProcessor.java:119]: There should be validation of the size of the archive before open it.

File: iac-extensions/kubernetes/src/main/java/org/sonar/iac/helm/utils/HelmFilesystemUtils.java [hunk 9]
```diff
@@ -111,4 +108,32 @@ private static String resolveToInputFile(Path helmDirectoryPath, InputFile addit
...
+
+  public static String normalizeToRuntimePathSeparator(String filename) {
+    return filename.replace('\\', File.separatorChar)
+      .replace('/', File.separatorChar);
+  }
+
+  public static boolean includeFile(String name) {
+    var normalized = normalizeToRuntimePathSeparator(name);
+    if (INCLUDED_EXTENSIONS.stream().anyMatch(normalized::endsWith)) {
+      var parent = new File(".");
```
Comment (mstachniuk) [inline iac-extensions/kubernetes/src/main/java/org/sonar/iac/helm/utils/HelmFilesystemUtils.java:120]: Instead of using `.` (current directory), the location of the tar file should be used.

Comment (mstachniuk) [reply iac-extensions/kubernetes/src/main/java/org/sonar/iac/helm/utils/HelmFilesystemUtils.java:120]: Or alternatively checking for following patterns in filename: `../`, `/..`, `..\`, `\..`, `~`, `./`, `.\`
